Adding Condition lab and Templates (#148)
aws-samples/cfn101-workshop: workshop/content/intermediate/templates/conditions/_index.md

* Identify sample use cases for leveraging Condition functions.
* Provision resources based on Condition evaluation.
* Conditionally specifying resource property values using Condition functions.

Let’s walk through an example of how to Conditions function in your template.

### Start Lab

#### **Defining Conditions at the resource level**

* Change directory to: `code/workspace/conditions`.
* Open the `condition-resource.yaml` template.
* Update the content of the template as you follow along steps on this lab.

Let’s get started!

Let’s start with specifying input parameters you want to use to make your template more reusable. You will add a `Parameters` section in your template, with an input parameter for your life cycle environment; you will call the parameter `EnvType`, and you will describe two example environment names, `test` and `prod`, as input values you allow. Define also an input parameter for the [Amazon Machine Image](https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/AMIs.html) (AMI) you will use: in this example, you use the example `LatestAmiId` parameter to refer, in its value, to the latest available Amazon Linux AMI.

{{% notice note %}}

> For more information about fetching latest AMI, see [Query for the latest Amazon Linux AMI IDs using AWS Systems Manager Parameter Store](https://aws.amazon.com/blogs/compute/query-for-the-latest-amazon-linux-ami-ids-using-aws-systems-manager-parameter-store/).

Reviewer: ```suggestion
For more information, see [Query for the latest Amazon Linux AMI IDs using AWS Systems Manager Parameter Store](https://aws.amazon.com/blogs/compute/query-for-the-latest-amazon-linux-ami-ids-using-aws-systems-manager-parameter-store/).
```

![condition-test](conditions/condition-test.png)

In the next step, you will create a new stack with the same template, and pass `prod` as the new value for the `EnvType` parameter, and verify that you will provision, with CloudFormation, your `Volume` and `MountPoint` resources. Navigate to the AWS CloudFormation [console](https://console.aws.amazon.com/cloudformation), and choose to create a stack using this template:

* In the CloudFormation console, select **Create stack With new resources (standard)**.
* In **Prepare template**, select **Template is ready**.
* In **Template source**, select **Upload a template file**.
* Choose the `condition-resource.yaml` template.
* Enter a **Stack name**. For example, choose to specify `cfn-workshop-condition-prod`.
* Pass `prod` as the value for the `EnvType` parameter. Choose **Next**.
* You can leave other Configure stack options default, Choose **Next**.
* Choose **Create stack**. You can view the progress of the stack being created in the CloudFormation console.
* Wait until the stack creation is complete. Refresh the view in the console until you see your stack to be in the `CREATE_COMPLETE` status.

> This time,  the `IsProduction` condition is true. Navigate to the **Resources** tab for your stack, and verify that along with your EC2 instance resource, your other `Volume` and `MountPoint` resources are also provisioned:

Reviewer: ```suggestion
This time, the `IsProduction` condition is true. Navigate to the **Resources** tab for your stack, and verify that along with your EC2 instance resource, your other `Volume` and `MountPoint` resources are also provisioned:
```

Author: Changed

Reviewer: @satyamgptaws I do not see a commit?

* You can leave other Configure stack options default, Choose **Next**.
* Choose **Create stack**. You can view the progress of the stack being created in the CloudFormation console.
* Wait until the stack creation is complete. Refresh the view in the console until you see your stack to be in the `CREATE_COMPLETE` status.

Once the Stack is in the `CREATE_COMPLETE` status, navigate to the **Resources** tab for your stack, and verify your EC2 instance is in the `CREATE_COMPLETE` status.

Next, verify the instance type is the one you expect: follow the link for the Physical ID of your instance, that will bring you to the Amazon EC2 Console:
![condition-test-property](conditions/condition-test-property.png)

You should see a view, as in the example shown next, showing that your instance type is `t2.micro`:
![ec2-instance](conditions/ec2-instance.png)

Similarly, you can run the same solution by passing the value of `EnvType` parameter as `prod`, and you will observe the instance type for provisioned EC2 instance will be `t2.small`.

Congratulations! You have learned how to conditionally specify resource property types!

### **Challenge**

> So far, you’ve learned how to use conditions with resources in your CloudFormation templates . Now let’s say you want to conditionally add an output in the [Outputs](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/outputs-section-structure.html) section of your `condition-resource.yaml` CloudFormation template.

Reviewer: ```suggestion
So far, you’ve learned how to use conditions with resources and property values in your CloudFormation templates. In this challenge, you will conditionally create an output in the [Outputs](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/outputs-section-structure.html) section of your `condition-resource.yaml` CloudFormation template.
```

> **Task:** Describe an [Outputs](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/outputs-section-structure.html) section in your template. Specify a Logical ID as `VolumeID`, and use the `Ref` [intrinsic function](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-ec2-instance.html#aws-properties-ec2-instance-return-values) to return the ID of your Volume. Now, update your existing CloudFormation Stack `cfn-workshop-condition-prod` with modified template to create an output based on `IsProduction` condition evaluation.

Reviewer: ```suggestion
**Task:** Describe an [Outputs](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/outputs-section-structure.html) section in your `condition-resource.yaml` template. Specify `VolumeId` for the Logical ID of your output, and use the `Ref` intrinsic function to [return the ID of your volume](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-ec2-instance.html#aws-properties-ec2-instance-return-values). Your goal, in this challenge, is to create your output only if the `IsProduction` condition is true: how can you reflect this intent in your template? Once ready, update your existing `cfn-workshop-condition-prod` stack with the template you updated, and verify your changes led to creating the output as you expected.
```

{{%expand "Need a hint?" %}}

* See the documentation for [Stack Output](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/outputs-section-structure.html#outputs-section-structure-examples), and define a `VolumeId` output in your template.

> * How do you conditionally create an output?

Reviewer: ```suggestion
* Review the documentation on [Conditions](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/conditions-section-structure.html) and [Associating a condition](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/intrinsic-function-reference-conditions.html#associating-a-condition). How do you conditionally create an output?
```

{{% /expand %}}

{{%expand "Want to see the solution?" %}}

Append the following content to the `condition-resource.yaml` file:

```yaml
Outputs:

>    VolumeID:

Reviewer: ```suggestion
   VolumeId:
```

      Value: !Ref Volume
      Condition: IsProduction
```

> Next lets update your stack  `cfn-workshop-condition-prod`, navigate to the AWS CloudFormation [console](https://console.aws.amazon.com/cloudformation), and choose to  `cfn-workshop-condition-prod` stack:

Reviewer: ```suggestion
Next, navigate to the AWS CloudFormation [console](https://console.aws.amazon.com/cloudformation), and choose to  update your `cfn-workshop-condition-prod` stack:
```

* In the CloudFormation console, select **Update stack**.

> * In **Prepare template**, select Replace current template **option**.

Reviewer: ```suggestion
* In **Prepare template**, select **Replace current template**.
```

* In **Template source**, select **Upload a template file**.
* Choose the `condition-resource.yaml` template.

> * You can leave the value of `EnvType` parameter as `prod`. Choose **Next**.

Reviewer: ```suggestion
* `EnvType` should already be set to `prod`. Choose **Next**.
```

* You can leave other Configure stack options default, Choose **Next**.
* Choose **Update stack**. You can view the progress of the stack being created in the CloudFormation console.
* Wait until the stack creation is complete. Refresh the view in the console until you see your stack to be in the `UPDATE_COMPLETE` status.

> Browse to the `Outputs` section of the stack, and validate `VolumeID` appears as an output:

Reviewer: ```suggestion
Navigate to the `Outputs` section of your stack, and validate the `VolumeId` output is present.
```

> ![condition-prod-update](conditions/condition-prod-update.png)

Reviewer: Please replace with a new image to reflect `ImageId` instead of `ImageID`.

Author: In the image, Outputs value shows VolumeID not ImageID.

{{% /expand %}}

> Note: content shown above is also available in relevant files located in the `code/solutions/conditions/condition-output.yaml` directory.

Reviewer: ```suggestion
The full solution is also available in the `code/solutions/conditions/condition-output.yaml` template file.
```

### Cleanup

> Follow these steps to clean up created resources:

Reviewer: ```suggestion
Follow steps shown next to clean up resources you created in this lab:
```

> * In the CloudFormation console, select the stack you have created in this lab. For example: `cfn-workshop-condition-test` .

Reviewer: ```suggestion
* In the CloudFormation console, choose the `cfn-workshop-condition-test` stack you created in this lab.
```

> * Choose **Delete** to delete the stack you created in this lab, and then **Delete stack** to confirm.

Reviewer: ```suggestion
* Choose **Delete** to delete the stack you created in this lab, and then choose **Delete stack** to confirm.
```

> Perform same steps as above for your other stacks as well: `cfn-workshop-condition-prod`, `cfn-workshop-condition-property-test`.

Reviewer: ```suggestion
Perform the same cleanup steps above for your other stacks you created in this lab: `cfn-workshop-condition-prod`, and `cfn-workshop-condition-property-test`.
```

---
### Conclusion

> Congratulations! You have learned how to conditionally create resources, and how to conditionally specify resource property types. For more information, see [Conditions](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/conditions-section-structure.html) and [Condition functions](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/intrinsic-function-reference-conditions.html).

Reviewer: ```suggestion
Congratulations! You have learned how to conditionally create resources, and how to conditionally specify resource property values. For more information, see [Conditions](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/conditions-section-structure.html) and [Condition functions](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/intrinsic-function-reference-conditions.html).
```
